src/software/util/design_patterns/constructor_arg_generic_factory_test.cpp

```diff
@@ -0,0 +1,95 @@
+#include "software/util/design_patterns/constructor_arg_generic_factory.h"
+
+#include <gtest/gtest.h>
+
+#include <exception>
+#include <iostream>
+
+// Create and register two test generics with the factory here
+class TestGeneric
+{
+   public:
+    virtual std::string info() = 0;
+};
+
+class TestGenericStringArg : public TestGeneric
+{
+   public:
+    TestGenericStringArg(std::string _text) : text(_text) {};
+    std::string info() override {return text;}
+    std::string text;
+};
+
+class TestGenericIntArg : public TestGeneric
+{
+   public:
+    TestGenericIntArg(int _number) : number(_number) {};
+    std::string info() override {return std::to_string(number);}
+    int number;
+};
+
+class TestGenericTwoArgs : public TestGeneric
+{
+   public:
+    TestGenericTwoArgs(int _number, std::string _string) : number(_number), string(_string) {
+        std::cout << number << " " << string << std::endl;
+    };
+    std::string info() override {return std::to_string(number) + ' ' + string;}
+    int number;
+    std::string string;
+};
+
+static TConstructorArgGenericFactory<std::string, TestGeneric, TestGenericStringArg, std::string> testFactoryA;
+static TConstructorArgGenericFactory<std::string, TestGeneric, TestGenericIntArg, int> testFactoryB;
+static TMultipleConstructorArgGenericFactory<std::string, TestGeneric,
+TestGenericTwoArgs(int, std::string)> testFactoryC;
+
+
+
+
+TEST(ConstructorArgGenericFactoryTest, test_create_generic_with_invalid_name)
+{
+    EXPECT_THROW((ConstructorArgGenericFactory<std::string, TestGeneric>::create("_FooBar_", 5)),
+                 std::invalid_argument);
+}
+
+TEST(ConstructorArgGenericFactoryTest, test_create_generic_with_invalid_arg_type)
+{
+    EXPECT_THROW((ConstructorArgGenericFactory<std::string, TestGeneric>::create(
+        "TestGenericStringArg", std::any(5))),
+                 std::invalid_argument);
+}
+
+TEST(ConstructorArgGenericFactoryTest, test_create_generic_with_valid_name_and_arg_type)
+{
+    auto type_ptr = ConstructorArgGenericFactory<std::string, TestGeneric>::create(
+        "TestGenericIntArg", 5);
+
+    EXPECT_TRUE(type_ptr);
+    EXPECT_EQ(static_cast<TestGenericIntArg*>(type_ptr.get())->number, 5);
+}
+
+TEST(ConstructorArgGenericFactoryTest, test_create_generic_with_two_args)
+{
+    auto type_ptr = ConstructorArgGenericFactory<std::string, TestGeneric>::create(
+        "TestGenericTwoArgs", 5, std::string("five"));
```

Comment:
I hope I'm wrong, but I thought of something that might cause this solution not to work. With the way we use factories currently, we would pass in the `IndexType` (e.g. "TestGenericTwoArgs") to the `ConstructorArgGenericFactory` and we don't know the types of the arguments are needed to construct the `TypeToCreate`. 

For example, we have an `IntStringTestGenericTwoArgs` that takes `int` and `string` as in the example above and an `IntIntTestGenericTwoArgs` that takes `int` and `int` as arguments. We also have a `type_string` contains either "IntStringTestGenericTwoArgs" or "IntIntTestGenericTwoArgs". Then, we run
```   
auto type_ptr = ConstructorArgGenericFactory<std::string, TestGeneric>::create(
        type_string, 5, std::string("five"));
```
If `type_string` contains "IntStringTestGenericTwoArgs", then I think this will work. However, if `type_string` contains "IntIntTestGenericTwoArgs", then I think this will throw an exception.

Perhaps all subtypes of a class need to share the same constructor args?

Reply:
Added an overload of `create` that accepts `IndexType` and `std::any`, where the `std::any` should contain a tuple of the constructor args